### PR TITLE
VACMS-12592: Checks that forms migration file isn't old

### LINKED ISSUE #12592: Add test for va forms csv to detect date of file
## Description

Similar to https://github.com/department-of-veterans-affairs/va.gov-cms/issues/10735 and the tests created there.  It just needs another test.

### Engineering notes / background
There is an @todo in the existing test file for where to add this test.  https://github.com/department-of-veterans-affairs/va.gov-cms/pull/12421/files#diff-87e4067ad8893e01fedfbf7ceee85fc5c4d63147b6f7cdd36f03f7a82373ea7fR18

### Tasks
Assuming we can check the timestamp, and the file is updated nightly (TODO: Verify based on comments below). 
* Confirm that the Forms CSV import machine (Jenkins?) can read timestamp from the Forms DB CSV hosting machine 
* Write the test - Check the timestamp is not > 3 days before today
* Find a way to hack proving it works with a local file that's > 3 days old

## Acceptance criteria

- [ ] If the date of the csv file that we move is over 3(? )days old a failure should be noted and prevent migration.  Three days should help avoid getting a false alarm Sunday evening if the file was not updated Saturday and Sunday.   This assumes it is updated m-f including holidays.

### Team
Please check the team(s) that will do this work.

- [ ] `CMS Team`
- [X] `Public Websites`
- [ ] `Facilities`
- [ ] `User support`

### PR DESCRIPTION
**Pull request by chri5tia**
## Description

Closes #12592

## Testing done
Not yet. Going to try running nightly tasks on the tugboat instance.

## Screenshots
n/a

## QA steps

TBD

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`

### Is this PR blocked by another PR?

- [ ] `DO NOT MERGE`

### Does this PR need review from a Product Owner

- [ ] `Needs PO review`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: scripts/tasks/VACMS-10735-Forms-CSV-migration-validation.php [hunk 1]
```diff
@@ -68,6 +67,18 @@ function _va_forms_validate_csv_headers(array $headers, string $exitMessage) {
  */
 function _va_forms_validate_csv_has_content(bool|array $csv, string $exitMessage) {
   if ($csv === FALSE) {
-    exit("{$exitMessage} Empty or invalid CSV. In _va_forms_validate_csv_contents()\r\n");
```
Comment (chri5tia) [inline scripts/tasks/VACMS-10735-Forms-CSV-migration-validation.php:71]: I am not actually sure if this change makes sense or not. I noticed that the exit message function doesn't match anything that already exists.

**Comment by ndouglas** [comment]
Does the command downloading the file update the last-modified date of the file to match the last-modified response header from downloading the file?  I don't understand how this works.

**Comment by chri5tia** [comment]
> Does the command downloading the file update the last-modified date of the file to match the last-modified response header from downloading the file?

It's not entirely clear because I haven't ssh'd into the actual server to look at the file. Not sure how to get that kind of access. But each time the file is modified/created on our end, it should have a unix modification date. This only checks that the file on the forms migration side was successfully updated. I am not sure if there is value in checking the age of the file on the source end. Does this help at all?

**Comment by ndouglas** [comment]
Sorry -- I was under the impression that you could curl that URL from Tugboat.  You can't.  It can only be accessed from the CMS, apparently, which is absolutely baffling but life goes on.

For future reference, see [here](https://github.com/department-of-veterans-affairs/va.gov-cms/blob/main/READMES/tugboat.md#tugboats-cli-tool-for-software-engineers) for info about shelling into Tugboat.

I see that the instructions say to check the last-modified date of the local file.  I'm not sure that's valuable, though, since that command will create a file even if the server responds with an error code:

```
$ curl https://www.va.gov/something-that-doesnt-exist -L -vvvv
...
> GET /something-that-doesnt-exist HTTP/1.1
...
< HTTP/1.1 404 Not Found
...
$ curl https://www.va.gov/something-that-doesnt-exist -L --output va_forms_data.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 43416    0 43416    0     0  55773      0 --:--:-- --:--:-- --:--:-- 56238
$ ls -lha va_forms_data.csv 
-rw-r--r-- 1 nathan.douglas staff 43K Mar 14 12:04 va_forms_data.csv
```

It seems that, so long as the server can be looked up via DNS and responds to the request in any way whatsoever, that the file will be created locally with a last-modified time of whenever the command was executed.  So I see little value in checking the last-modified time of that file.

If, however, you shell into Tugboat, you find... that the URL is not curlable from Tugboat.  Presumably a security measure.  Ugh, and also that shouldn't surprise me at this point.  So if you shell into `staging.cms.va.gov` (which, don't worry, you won't need to), then curl the URL, you'll see:

```
sh-4.2$ curl https://vaww.webdevi.va.gov/vaforms/VAForms_DataExtract/VAForms_FormsData.txt -vvvv
...
> GET /vaforms/VAForms_DataExtract/VAForms_FormsData.txt HTTP/1.1
...
< HTTP/1.1 200 OK
< Content-Type: text/plain
< Last-Modified: Tue, 14 Mar 2023 03:30:02 GMT
< Accept-Ranges: bytes
< ETag: "f39b54432556d91:0"
< Server: Microsoft-IIS/8.5
< X-Powered-By: ASP.NET
< Date: Tue, 14 Mar 2023 16:09:10 GMT
< Content-Length: 531846
< 
```

That Last-Modified response header is probably what we should be checking.  That indicates the last-modified time of the data on the server, and is at present apparently reliable.  I wouldn't trust it longterm.  There's no reason why that couldn't read a date in the distant past, or the future, or the birthday of a Very Special DevOps Engineer.

Nevertheless, the issue is open.  Personally, I would use PHP's CURL interface to check the same URL to validate the headers.  And, frankly, I'd move the whole thing into PHP and locate it within a Drush command because _then_ we can use dependency injection to actually be able to automate tests for the command, since we're apparently not able to test this functionality either locally or even on Tugboat, even manually, as it is currently written.

Obviously that's a total rewrite of the ticket.  So I don't know.  I'm not anybody's boss here, but in my opinion this ticket as written does not do a useful thing.  And the only way to make it useful is to do a lot more work.

To just close out the ticket, I would suggest using CURL to check the Last-Modified header of the URL.  I would pass the URL in as an argument or an environment variable to the script so that we DRY.

This is all open to debate.  Maybe I'm getting something wrong.

**Comment by chri5tia** [comment]
Thank you for the thorough input. Can you explain why simply checking the unix date modified of the csv isn't simple and elegant enough? It seems like there are a lot of ways to make sure a file isn't old, but this way only required a couple lines of code. Not questioning your wisdom, just want more understanding about it e.g. why the header makes more sense, etc.

**Comment by chri5tia** [comment]
So I was wondering if this test was supposed to make sure that a new file got grabbed. I know that the file grabbed will have a new date, but if something goes wrong with grabbing a file and the old one is still sitting on the server, then it would have an old date. I don't think we need to make sure that the source has changed or not. Thoughts?

**Comment by ndouglas** [comment]
To be clear, we have this code in `tasks-periodic.yml`:

```
  # Run with: ./bin/task --taskfile=tasks-periodic.yml daily
  # http://jenkins.vfs.va.gov/job/cms/job/cms-daily
  daily:
    cmds:
      - |
        cat <<EOF | bash
          exit_code=0
          tasks=(
            va/background/daily/migrate_copy/va_forms_csv_source
            va/background/daily/migrate/va_forms
            va/background/daily/update/taxonomy_entity_index
          )
          trap '{ (( exit_code |=\$? )); }' ERR
          for i in "\${tasks[@]}"; do
            task --taskfile=./tasks-periodic.yml \$i
          done
          trap - ERR
          exit \$exit_code;
        EOF
```

That runs, executes each of the `tasks` items individually, and returns a failing error code if any of the tasks fail.

The two tasks we care about right now are these:

```

  va/background/daily/migrate_copy/va_forms_csv_source:
    desc: This is because we cannot access the server on the SOCKS proxy right now.
    cmds:
      - |
        cd docroot/sites/default/files/migrate_source
        curl https://vaww.webdevi.va.gov/vaforms/VAForms_DataExtract/VAForms_FormsData.txt -L --output va_forms_data.csv

  va/background/daily/migrate/va_forms:
    desc: Migrate va forms.
    cmds:
      - drush $DRUSH_ALIAS scr scripts/tasks/VACMS-10735-Forms-CSV-migration-validation.php
      - drush $DRUSH_ALIAS migrate:reset-status va_node_form
      - drush $DRUSH_ALIAS migrate:import va_node_form
```

(the `migrate_copy` should be `migrate`, and there's no need for these to be split into separate tasks, but one thing at a time)

`va/background/daily/migrate_copy/va_forms_csv_source` changes into the `docroot/sites/default/files/migrate_source` directory and runs the `curl` command.  If it fails, it fails... but `va/background/daily/migrate/va_forms` will run regardless.

So `va/background/daily/migrate/va_forms` might run with an empty file at `docroot/sites/default/files/migrate_source/va_forms_data.csv`.  It might run with HTML in that file.  It might run with _anything_. 

What do we know about that file?

We know the file exists now, on production.  See [here](https://prod.cms.va.gov/sites/default/files/migrate_source/va_forms_data.csv).

And if the file exists, and the `curl` command fails utterly, the `curl` command will leave it in place:

```
$ ls -lha va_forms_data.csv 
-rw-r--r-- 1 nathan.douglas staff 43K Mar 14 12:04 va_forms_data.csv
$ curl https://www.vawefwfwegwegewgwe.gov/something-that-doesnt-exist -L --output va_forms_data.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: www.vawefwfwegwegewgwe.gov
$ ls -lha va_forms_data.csv 
-rw-r--r-- 1 nathan.douglas staff 43K Mar 14 12:04 va_forms_data.csv
```

And if the `curl` command results in a status code that isn't a 200 -- like a 502, 404, etc -- the `curl` command will create the file anew, but it won't contain what we want:

```
$ ls -lha va_forms_data.csv 
-rw-r--r-- 1 nathan.douglas staff 43K Mar 14 12:04 va_forms_data.csv
$ curl https://www.va.gov/something-that-doesnt-exist -L --output va_forms_data.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 43416    0 43416    0     0  36272      0 --:--:--  0:00:01 --:--:-- 36453
$ ls -lha va_forms_data.csv 
-rw-r--r-- 1 nathan.douglas staff 43K Mar 14 13:09 va_forms_data.csv
```

So, yes, it is elegant.  But the code doesn't make any useful assertions about the quality of the downloaded file.  The only case in which it'll fail is if the server goes away completely and is down for longer than three days.  Only then will the local `va_forms_data.csv` stop being updated.  Only then will the check fail.

**Comment by chri5tia** [comment]
Ah okay. Maybe we need to bring this ticket back to the drawing board and move it into the next sprint. Would you be willing to meet with me to discuss this and redraw the user story and ACs?

**Comment by ndouglas** [comment]
>Would you be willing to meet with me to discuss this and redraw the user story and ACs?

Sure!  My calendar's fairly open.

**Comment by jilladams** [comment]
Chatted with Christia on this. She'll schedule time with Nate to review, and noting: this change came up out of PR review over here: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/12421, where there were tests already added to verify that the CSV isn't empty and has the expected headers, for reference.

**Pull request closed without merging**